[DOCS] Expands classification documentation with multiclass (#973)
elastic/stack-docs: docs/en/stack/ml/df-analytics/dfa-classification.asciidoc

experimental[]

{classification-cap} is a {ml} process that enables you to predict the class or
category of a data point in your data set. Typical examples of {classification}
problems are predicting loan risk, classifying music, or detecting cancer in a

> DNA sequence. In the first case, for example, the data set might contain the

Reviewer: Hello everyone! 👋 
Very minor issue, but I would consider rewording this sentence `detecting cancer in a
DNA sequence` to something like `detecting the potential for cancer from a DNA sequence` , because technically cancer is an illness in a tissue/organ and not in the DNA sequence.

When you create a {classification} job, you must specify which field contains 
the classes that you want to predict. This field is known as the _{depvar}_. It
must contain no more than 30 classes. By default, all other
{ref}/put-dfanalytics.html#dfa-supported-fields[supported fields] are included
in the analysis and are known as _{feature-vars}_. The runtime and resources
used by the job increase with the number of feature variables. Therefore, you
can optionally include or exclude fields from the analysis. For more information
about field selection, see the
{ref}/explain-dfanalytics.html[explain data frame analytics API].

[discrete]
[[dfa-classification-supervised]]
==== Training the {classification} model

{classification-cap} – just like {regression} – is a supervised {ml} process.
When you create the {dfanalytics-job}, you must provide a data set that contains
the _ground truth_. That is to say, your data set must contain the {depvar} 
and the {feature-vars} fields that are related to it. You can divide the data
set into training and testing data by specifying a `training_percent`. By
default when you use the

> {ref}/put-dfanalytics.html[create {dfanalytics-jobs} API], 100% of the eligible

Reviewer: Could we please clarify what `eligible` documents are? For ex, I know we exclude documents if they have a field that has an array with more than one element (cc: @dimitris-athanasiou - could you please assist in getting this worded correctly). We don't have to define it in this paragraph, but it would be nice to link to a section that explains what documents are eligible.

documents in the data set are used for training. If you divide your data set,
the job stratifies the data to ensure that both the training and testing data

> sets are representative of the full data set. That is to say, the same

Reviewer: I would clarify the concept of stratification here by saying `the training and testing data contains classes in proportions that are representative of the class proportions in the full data set` and maybe take out the next sentence - the one starting with `That is to say,` (cc: @tveasey )

proportions of each class are fairly represented in both data sets.

When you are collecting a data set to train your model, ensure that it
captures information for all of the classes. If some classes are poorly

> represented in the training data set, the model might be unaware of them. In

Reviewer: I would clarify this part `If some classes are poorly
represented in the training data set` by saying something like `If some classes are poorly
represented in the training data set (that is, you have very few data points per class)`

The effects of imbalanced data are automatically mitigated before the 
training. Nonetheless, it is a good idea to train your model with a data set 
that is approximately balanced. That is to say, ideally your data set should 
have a similar number of data points for each class.
////

[discrete]
[[dfa-classification-algorithm]]
===== {classification-cap} algorithms

//tag::classification-algorithms[]
{classanalysis-cap} uses an ensemble algorithm that is a type of boosting called 
boosted tree regression model which combines multiple weak models into a 
composite one. It uses decision trees to learn to predict the probability that a 
data point belongs to a certain class. A sequence of decision trees are trained 
and every decision tree learns from the mistakes of the previous one. Every tree 
is an iteration of the last one, hence it improves the decision made by the 
previous tree.
//end::classification-algorithms[]

[discrete]
[[dfa-classification-performance]]
==== {classification-cap} performance

As a rule of thumb, a {classanalysis} with many classes takes more time to run 
than a binary {classification} process when there are only two classes. The 

> relationship between the number of classes and the runtime is linear.

Reviewer: I would say the `runtime is roughly linear` here to allow ourselves some wiggle room.

The runtime also scales approximately linearly with the number of involved 

> documents below 200,000 data points. Therefore, if you double the number of 

Reviewer: I think this paragraph might be a too much detail, because the exact conditions that we have when testing the runtime won't necessarily translate to the user's environment, so they might experience something different. I would leave in this sentence starting with `Therefore, if you double the number of  documents, ...`
I would remove this `Above 200,000 
data points, runtime increases faster.`, because it is a bit too specific.

documents, then the runtime of the analysis doubles respectively. Above 200,000 
data points, runtime increases faster.

To improve the performance of your {classanalysis}, consider using a smaller 
`training_percent` value when you create the job. That is to say, use a smaller 

> percentage of your documents to train the model more quickly. If possible, 

Reviewer: I would add here that the key is iteration. The user can start by using a smaller training percentage, run the job, check the performance and then decide if it's necessary to increase the training percentage.

experimental[]

{classification-cap} is a {ml} process for predicting the class or category of a 
given data point in a data set. Typical examples of {classification} problems are 
predicting loan risk, classifying music, or detecting cancer in a DNA sequence. 
In the first case, for example, our data set consists of data on loan applicants 
that covers investment history, employment status, debit status, and so on. 
Based on historical data, the {classification} analysis predicts whether it is 
safe or risky to lend money to a given loan applicant. In the second case, the 
data we have represents songs and the analysis – based on the features of the 
data points – classifies the songs as hip-hop, country, classical, or any 
other genres available in the set of categories we have. Therefore, 
{classification} is for predicting discrete, categorical values, unlike 
{reganalysis} which predicts continuous, numerical values. The maximum number of 
classes for {classanalysis} is 30.

[discrete]
[[dfa-classification-depvar]]

> ==== {depvar-cap}

Reviewer 2: > {depvar-cap}

This attribute doesn't exist in the shared attributes file, so it's not resolving.  I actually recommend omitting this section header and anchor.  I think this paragraph flows well from the one above it.

When you create a {classification} job, you must specify which field contains 

> the classes that you want to predict. We refer to this field as {depvar}. The 

Reviewer 2: ```suggestion
the classes that you want to predict. This field is known as the {depvar}. The 
```

rest of the fields that are included in the analysis are used to create the 
model for predicting the value of the {depvar} (these are the {feature-vars}). 
All {ref}/put-dfanalytics.html#dfa-supported-fields[supported fields] are 
included by default, however, you can explicitly exclude and/or include fields. 

> Since the runtime and the necessary resources for the analysis are increasing as 

Reviewer 2: ```suggestion
Since the runtime and the necessary resources for the analysis increase as 
```

> the number of included fields is growing, consider to include only relevant 

Reviewer 2: ```suggestion
the number of included fields grow, consider including only relevant 
```

fields. For more information about field selection, see 
{ref}/explain-dfanalytics.html[Explain data frame analytics API].

[discrete]
[[dfa-classification-supervised]]
==== Training the {classification} model

{classification-cap} – just like {regression} – is a supervised {ml} process. It 
means that you need to supply a labeled training data set that has a {depvar} 
and some fields that are related to it. The {classification} algorithm learns 
the relationships between these fields and the {depvar}. Once you’ve trained the 
model on your training data set, you can reuse the knowledge that the model has 
learned about the relationships between the data points to classify new data.

> Although the effects of imbalanced data are automatically mitigated before the 

Reviewer 2: Not mandatory, but I think it would be more impactful to split this sentence.
```suggestion
The effects of imbalanced data are automatically mitigated before the 
```

> training, we still recommend using an approximately balanced data set to train 

Reviewer 2: Another minor suggestion to remove "we" if possible:
```suggestion
training. Nonetheless, it is a good idea to train your model with a data set that is approximately balanced.
```

> the model. It means the number of data points belonging to the various classes 

Reviewer 2: I find when sentences start with "it", I have to go back to previous sentences to figure out what "it" is.  So I'd reword to something like this:
```suggestion
That is to say, ideally your data set should have a similar number of data points for each class.
```

> is not widely different.

Reviewer 2: If you merge the previous suggestions, this sentence fragment is not longer required:
```suggestion

```

that is approximately balanced. That is to say, ideally your data set should 
have a similar number of data points for each class.

[discrete]
[[dfa-classification-algorithm]]
===== {classification-cap} algorithms

//tag::classification-algorithms[]
{classanalysis-cap} uses an ensemble algorithm that is a type of boosting called 
boosted tree regression model which combines multiple weak models into a 
composite one. It uses decision trees to learn to predict the probability that a 
data point belongs to a certain class. A sequence of decision trees are trained 
and every decision tree learns from the mistakes of the previous one. Every tree 
is an iteration of the last one, hence it improves the decision made by the 
previous tree.
//end::classification-algorithms[]

[discrete]
[[dfa-classification-performance]]
==== {classification-cap} performance

As a rule of thumb, a {classanalysis} with many classes takes more time to run 
than a binary {classification} process when there are only two classes. The 
relationship between the number of classes and the runtime is linear.

The runtime also scales approximately linearly with the number of involved 

> documents below 200.000 data points. Therefore, if you double the number of 

Reviewer 2: Our style guide recommends using a comma to separate out numbers larger than 999.
```suggestion
documents below 200,000 data points. Therefore, if you double the number of 
```

> documents, then the runtime of the analysis doubles respectively. Above 200.000 

Reviewer 2: Ditto re thousand separator:
```suggestion
documents, then the runtime of the analysis doubles respectively. Above 200,000 
```

data points, runtime increases faster.

> To improve the performance of your {classanalysis}, consider to set a smaller 

Reviewer 2: Minor wording change. I think it works better as "setting" or "using" than "to set".
```suggestion
To improve the performance of your {classanalysis}, consider using a smaller 
```

`training_percent` value when you create the job. That is to say, use a smaller 
percentage of your documents to train the model more quickly. If possible, 

> prepare your input data such that is has less classes. You can also remove the 

Reviewer 2: Small typo
```suggestion
prepare your input data such that it has less classes. You can also remove the 
```

{feature-vars-cap} are the values that the {depvar} value depends on. There are 
three different types of {feature-vars} that you can use with our 
{classification} algorithm: numerical, categorical, and boolean. Arrays are not 

> supported in the {feature-var} fields.  

Reviewer 2: > Feature variables

This can perhaps be pursued in a separate PR, but it occurred to me while reading this page that neither our UI nor the API refer to "feature variables". They refer to a "dependent variable" and to included or excluded "analyzed fields". Should we therefore remove this section label and term? Or make it less prominent?

I think this section boils down to something like this: When you create a classification job, you must specify which field contains the classification labels/values/categories that you want to predict (no more than 30). You can also optionally specify which fields to include or exclude from the analysis.  [It might be nice to mention here why you would choose to explicitly include/exclude fields as opposed to letting the analytics do it for you. Why bother? And also maybe mention how you can use the explain API to see which fields are used]

Author: @lcawl 
In 860f884, I made feature variables less prominent. I still use the term between brackets though, I felt it won't hurt to name it even if we don't use the term on the UI or in the API docs. If you think we should omit it entirely, we can easily delete it without further complications. Also mentioned why you would choose to include/exclude fields and mentioned the Explain API with a link. I also removed the details of supported fields, but put a link that points to the supported field-types in the PUT DFA API docs (which creates a dependency between this PR and the other creating the link).

{feature-vars} and a {depvar}. The {classification} algorithm learns the 
relationships between the features and the {depvar}. Once you’ve trained the 
model on your training dataset, you can reuse the knowledge that the model has 
learned about the relationships between the data points to classify new data. In 
case there are more than two classes, an automated process called stratification 
makes sure that the proportions of each class in the training data are equal so 
that the training is consistent. 

> Your training dataset should be approximately balanced. It means the number of 

Reviewer 2: Minor detail: "dataset" is two words in the style guide
```suggestion
Your training data set should be approximately balanced. It means the number of 
```

> The ensemble algorithm that we use in the {stack} for {classification} is a type 

Reviewer 2: > The ensemble algorithm that we use....

This can likely be addressed in a separate PR, but I think we can rewrite these sentence to take "us" out of the picture.  For example, something like:  Classification analysis uses an ensemble algorithm... It uses decision trees...

of boosting called boosted tree regression model which combines multiple weak 
models into a composite one. We use decision trees to learn to predict the 
probability that a data point belongs to a certain class. A sequence of decision 
trees are trained and every decision tree learns from the mistakes of the 
previous one. Every tree is an iteration of the last one, hence it improves the 
decision made by the previous tree.
//end::classification-algorithms[]

[discrete]
[[dfa-classification-performance]]
==== {classification-cap} performance

As a rule of thumb, a {classanalysis} with many classes takes more time to run 
than a binary {classification} process when there are only two classes. The 
relationship between the number of classes and the runtime is linear.

The runtime also scales approximately linearly with the number of involved 
documents below 200.000 data points. Therefore, if you double the number of 
documents, then the runtime of the analysis doubles respectively. Above 200.000 
data points, runtime increases faster.

> To improve the performance of your {classanalysis}, consider to set a smaller 

Reviewer 2: > To improve the performance of your {classanalysis}, consider to set a smaller training percent. 

When I read this, it's not immediately obvious what this means.  I had to go look at the API for a definition. Could we maybe change it to something like this?:

> To improve... consider using a smaller `training_percent` value when you create the job. That is to say, use a smaller percentage of your documents to train the model more quickly.

The ensemble algorithm that we use in the {stack} for {classification} is a type 
of boosting called boosted tree regression model which combines multiple weak 
models into a composite one. We use decision trees to learn to predict the 
probability that a data point belongs to a certain class. A sequence of decision 
trees are trained and every decision tree learns from the mistakes of the 
previous one. Every tree is an iteration of the last one, hence it improves the 
decision made by the previous tree.
//end::classification-algorithms[]

[discrete]
[[dfa-classification-performance]]
==== {classification-cap} performance

As a rule of thumb, a {classanalysis} with many classes takes more time to run 
than a binary {classification} process when there are only two classes. The 
relationship between the number of classes and the runtime is linear.

> The runtime scales approximately linearly with the number of involved documents 

Reviewer 2: ```suggestion
The runtime also scales approximately linearly with the number of involved documents 
```

below 200.000 data points. Therefore, if you double the number of documents, 
then the runtime of the analysis doubles respectively. Above 200.000 data 
points, runtime increases faster.

> To achieve faster runtime, consider to set a smaller training percent, prepare 

Reviewer 2: > To achieve faster runtime...

This sentence is similar to the previous sentence ("runtime increases faster"), but their meanings are different in a way that I find a bit confusing.  To avoid confusion, maybe we could change this sentence to something like:

> To improve the performance of your {classanalysis}....

> your input data to have less classes if it is possible, and remove the fields 

Reviewer 2: This sentence is quite long, so I'd recommend splitting it:
```suggestion
If possible, prepare your input data such that is has less classes. You can also remove the fields 
```

The ensemble algorithm that we use in the {stack} for {classification} is a type 
of boosting called boosted tree regression model which combines multiple weak 
models into a composite one. We use decision trees to learn to predict the 
probability that a data point belongs to a certain class. A sequence of decision 
trees are trained and every decision tree learns from the mistakes of the 
previous one. Every tree is an iteration of the last one, hence it improves the 
decision made by the previous tree.
//end::classification-algorithms[]

[discrete]
[[dfa-classification-performance]]
==== {classification-cap} performance

> As a rule of thumb, a {classanalysis} with many classes takes more time to run 

Reviewer 2: > As a rule of thumb...

This sentence seems like it should be at the start of the next paragraph, not this one. This paragraph seems to be talking mostly about the number of documents, whereas the next one is talking about the number of classes.

[[dfa-classification-evaluation]]
==== Measuring model performance

You can measure how well the model has performed on your data set by using the 
`classification` evaluation type of the 
{ref}/evaluate-dfanalytics.html[evaluate {dfanalytics} API]. The metric that the 
evaluation provides you is a confusion matrix. The more classes you have, the 
more complex the confusion matrix is. The matrix tells you how many data points 
that belong to a given class were classified correctly and incorrectly. In other 
words, how many data points that belong to the X class were classified correctly 
and mistakenly classified as Y or Z or any other class in the data set other than 

> X.

Reviewer 2: > In other words, how many data points...

In my opinion, this sentence is redundant (i.e. doesn't add anything to the previous statement). What I think would be more interesting to add (in a separate PR if necessary) is what to do about poor classification results.  Can we provide tips about whether to improve it by re-running with feature importance enabled or using the explain API to determine which fields we should explicitly include or exclude? Or by curating the training data?

Author: I deleted the sentence and address the rest in a separate PR.
